Fix range operator

iSazonov opened the pull request:
## PR Summary
Fix range operator enumerator. Close #5519
Fix [char] cast in range operator. Close #5731

Breaking-change:  "0".."9" returns [char] previously in PSCore, now it returns [int]. After the change, the behavior is the same as in Windows PowerShell.

## PR Checklist

Note: Please mark anything not applicable to this PR `NA`.

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [x] Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] User facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [NA] Issue filed - Issue link:
- [ ] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
    - [NA] [Add `[feature]` if the change is significant or affectes feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.

iSazonov committed:
Fix range operator enumerator
Fix range operator enumerator
Fix [char] cast in range operator
Add tests

markekraus commented on src/System.Management.Automation/engine/lang/parserutils.cs:
```diff
@@ -1521,6 +1527,81 @@ public bool MoveNext()
+
+        private int _current;
+
+        Object IEnumerator.Current
```
Public properties should go before Internal properties https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1202.md

markekraus commented on src/System.Management.Automation/engine/lang/parserutils.cs:
```diff
@@ -1521,6 +1527,81 @@ public bool MoveNext()
+            get { return Current; }
+        }
+
+        public string Current
```
Public properties should go before Internal properties https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1202.md

markekraus commented on src/System.Management.Automation/engine/lang/parserutils.cs:
```diff
@@ -1521,6 +1527,81 @@ public bool MoveNext()
+            get { return _current; }
+        }
+
+        private int _increment = 1;
```
Fields not backing public properties should be defined before properties . https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1202.md

iSazonov replied:
@markekraus I reformatted the class.

markekraus commented on src/System.Management.Automation/engine/lang/parserutils.cs:
```diff
@@ -1478,7 +1478,13 @@ internal int UpperBound
         private int _current;
-        public object Current
+
+        Object IEnumerator.Current
```
This should swap places with `private int _current;` so that the field can be closed to the property it backs.

markekraus commented on src/System.Management.Automation/engine/lang/parserutils.cs:
```diff
@@ -1521,6 +1527,81 @@ public bool MoveNext()
+    /// The simple enumerator class is used for the range operator '..'
+    /// in expressions like 'A'..'B' | ForEach-Object { $_ }
+    /// </summary>
+    internal class CharRangeEnumerator : IEnumerator
```
I'm curious if we should start following the one-type-per-file rule https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1402.md

lzybkr replied:
It is often useful to have related types in the same file, so I'm strongly opposed to following this rule. I'm strongly in favor of organizing code in ways that make sense to the author and reviewers. We shouldn't blindly follow the guidelines of some tool.

markekraus replied:
I agree that it make sense to group related code and that following the rules blindly is silly. But, some of the files in this project are so massive that there is 0 benefit gained by the code being grouped, IMO. If I need to have 3 tabs open on the same file to work on it effectively, the code is probably better split, IMO. Doesn't have to be strictly followed, but, maybe give some serious consideration?

lzybkr replied:
Are you suggesting the author of said code saw 0 benefit of a huge file, because you might be saying that to said author. (I'm the primary author of 4 of the 10 largest non-generated files in the repo.)

To pick on the largest file - Ast.cs - there are at least 72 classes in that file. When I know I'm looking for something Ast related, but not sure what, if I end up with 72 open tabs that I need to close, I'm less productive than if I open 1 file.

And if you needed 3 tabs open to effectively work on some code, you needed 3 tabs to see some code, 1 file or 3 files, right? I do realize 

There is always room for improvement, but it should be thoughtful and not rely on some tool - tools are good at enforcing things that are easy to check, less good at providing good judgement.

markekraus replied:
>Are you suggesting the author of said code saw 0 benefit of a huge file,

Not at all. I'm suggesting *I* see 0 benefit.

> because you might be saying that to said author.

I know 🙂 .

> And if you needed 3 tabs open to effectively work on some code, you needed 3 tabs to see some code, 1 file or 3 files, right?

Right, I'm saying I see no benefit to these being in the same file, where I do see benefit in it being in separate files. I think it's a difference of preference, I prefer smaller files grouped in folders over monolith files. It's easier to jump between code in different files using the IDE, IMO. When the code is in the same file, the default action is to jump to the code in the current tab. When it is in another file, a new tab opens with that file and my current place in the previous file is preserved. *shrugs

I'm also trying to approach this from a maintainable open source project perspective. If this rule doen't makes sense for this project, then it doesn't make sense and we don't use it. I just want to have something in place that can help maintain and encourage an enforceable style for the life of the project, preferably through CI. It can help shift the burden from reviewers to contributors. 

If I'm wasting my efforts there, then I will gladly redirect my attention elsewhere.

markekraus reviewed (approved):
Style comments only. Not blocking.

lzybkr commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -4295,9 +4298,18 @@ private Expression GetRangeEnumerator(ExpressionAst condExpr)
-                    result = Expression.New(CachedReflectionInfo.RangeEnumerator_ctor,
-                                            lhs.Convert(typeof(int)),
-                                            rhs.Convert(typeof(int)));
+                    if (lhs.Type == typeof(string) || lhs.Type == typeof(Char))
```
This doesn't support variables, e.g.
```powershell
$a = 'a'
$z = 'z'
$a..$z
```

iSazonov replied:
Added support variables.

iSazonov committed:
Add support variables in range operator

iSazonov committed:
Reformat CharRangeEnumerator

iSazonov commented:
MacOS CI temporary failed.

markekraus commented on src/System.Management.Automation/engine/lang/parserutils.cs:
```diff
@@ -1521,6 +1556,81 @@ public bool MoveNext()
+
+        private bool _firstElement = true;
+
+        public CharRangeEnumerator(char lowerBound, char upperBound)
```
This should probably go with the other methods in the class.

iSazonov replied:
I use your link https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1201.md
 - constructors is after fields.

markekraus replied:
Ah I missed that this was the constructor... My bad

markekraus commented on src/System.Management.Automation/engine/lang/parserutils.cs:
```diff
@@ -1521,6 +1556,81 @@ public bool MoveNext()
+    /// The simple enumerator class is used for the range operator '..'
+    /// in expressions like 'A'..'B' | ForEach-Object { $_ }
+    /// </summary>
+    internal class CharRangeEnumerator : IEnumerator
```
Should this implement `IEnumerable<String>` instead?

iSazonov replied:
I think we cast all to `object` type before send to pipe. So it seems there is no sense to specify here the exact type.

markekraus commented:
@iSazonov I restarted the macOS job for you.

vors commented:
On this topic, it would be nice, if we can utilize new C# Slice syntax someday for perf reasons.

lzybkr commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -4295,12 +4299,24 @@ private Expression GetRangeEnumerator(ExpressionAst condExpr)
-                    result = Expression.New(CachedReflectionInfo.RangeEnumerator_ctor,
-                                            lhs.Convert(typeof(int)),
-                                            rhs.Convert(typeof(int)));
+                    if ((lhs.Type == typeof(string) || lhs.Type == typeof(char))
```
I think this will break a weird but supported scenario:
```powershell
"0".."10"
```
I know it's dumb code, but I've seen it often enough that I wouldn't break this case.

iSazonov replied:
😄 I wonder why do we have CIs if we have Jason Shirk?

While I don't know how fix this here. And I see the same problem in line 4893. I guess we can not resolve the conflict in `RangeOperator` method: "0".."9", $var1.."9", "0"..$var2 - covert string to char or int? What if user want get char array? 😕 

Another question. In line 4893 I tried to make this code faster for constants. It seems that with this addition it can becomes too complicated and duplicated. I think to leave only Expression.Call() and put all magic code in one method. Right?
I see all `Expression.Call()` has a ToDo comment ("replace this with faster code") - could you please explain the idea? Maybe we can make more right thing now?

lzybkr replied:
Assume any string with Length > 1 or if the only character is 0-9, assume int, otherwise char - that should be sufficient.

The `TODO` was written when I first wrote that code and just getting things working. I don't remember what I had in mind, if anything.

It's your call on how you want to write the code - you can avoid some duplication with a helper method if you prefer to keep the optimization.

iSazonov committed:
Add support for special case "0".."10"

iSazonov commented:
@lzybkr @SteveL-MSFT Could you please review?

SteveL-MSFT reviewed (changes requested):
Looks good.  However, I think for the letter range tests, it would be better to go A..C to ensure B gets generated rather than just the two ends for A..B

iSazonov committed:
Extend char ranges in tests

lzybkr commented on src/System.Management.Automation/engine/lang/parserutils.cs:
```diff
@@ -785,6 +785,94 @@ internal static object JoinOperator(ExecutionContext context, IScriptExtent erro
+            //     [char]'A'..[char]'Z'
+            // Special case:
+            //     [char]'0'..[char]'9'
+            if (lType == typeof(char) && rType == typeof(char))
```
You can avoid the calls to GetType() and casts by using C#7 pattern matching like this:
```C#
if (lbase as char lc && rbase as char rc)
{
    return CharOps.Range(lc, rc);
}
```

The same would apply below for strings.

iSazonov replied:
Fixed.

lzybkr commented on src/System.Management.Automation/engine/lang/parserutils.cs:
```diff
@@ -785,6 +785,94 @@ internal static object JoinOperator(ExecutionContext context, IScriptExtent erro
+            // For arguments of string type and length == 1
+            // in the special case:
+            //     "0".."9" - return objects of [int] type
+            if (lType == typeof(string) && rType == typeof(string))
```
It might nice to support:

```
[char]"a" .. "z"
```

It would be less useful to have `"a"..[char]"z"`, but it's not complicated to support both.

You could introduce a local helper function:
```c#
object AsChar(object o)
{
    if (o is char) return o;
    if (o as string s && s.Length == 1 && !Char.IsDigit(s, 0)) return o[0];
    return null;
}

if (AsChar(lbase) as char lc && AsChar(rbase) as char rc)
{
    return CharOps.Range(lc, rc);
}
```

This covers all combos of char/char,string/string, and mixed. I think - browser compiled so may not work :)

iSazonov replied:
Thanks for great comment!

Done.

lzybkr commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -4874,14 +4871,11 @@ public object VisitBinaryExpression(BinaryExpressionAst binaryExpressionAst)
-                    return Expression.Call(CachedReflectionInfo.IntOps_Range,
-                                           lhs.Convert(typeof(int)),
-                                           rhs.Convert(typeof(int)));
+                    // TODO: replace this with faster code
```
The idea behind "faster" is to avoid extra runtime type checks.

The general pattern using `Expression.Dynamic` is to generate code like (for a binary operator like this):

```
if (lhs.GetType() == int && rhs.GetType() == int))
    return IntOps.Range((int)lhs, (int)rhs);
Find some other binding
```

If there is no binding that works, the binder generates a new site, perhaps to throw an error, or add conversions, or whatever. But the key thing here is there are no extra checks in the ideal scenario, if the lhs and rhs don't change types the next time you execute the same line in script, you have exactly the code you need.

The assumption is types don't change much, but you need to generate code that can handle that.

At any rate, it's probably not that important here, but I thought I'd expand on why the code can be made faster.

iSazonov replied:
Thanks for the deep explaination!

So if we cast all to `object` here we generate the code only one time? And now we delegate type checks to `AsChar()`. If we could move these checks from the function to this level (line 4875), then we could generate more specific sites and be faster, but I don't know if it is possible and reasonable.

lzybkr replied:
Definitely possible and should be faster, but it might only be worthwhile as an exercise in learning how `Expression.Dynamic` and call sites work.

lzybkr commented on src/System.Management.Automation/engine/lang/parserutils.cs:
```diff
@@ -785,6 +785,94 @@ internal static object JoinOperator(ExecutionContext context, IScriptExtent erro
+            var lType = lbase.GetType();
+            var rType = rbase.GetType();
+
+            if (lType == typeof(char) && rType == typeof(char))
```
My suggestions for C#7 pattern matching apply here as well.

iSazonov replied:
Fixed.

lzybkr commented on src/System.Management.Automation/engine/lang/parserutils.cs:
```diff
@@ -1521,6 +1615,81 @@ public bool MoveNext()
+
+        internal int LowerBound
+        {
+            get { return _lowerBound; }
```
Can you use an auto-property instead?
```
internal int LowerBound
{
    get; private set;
}
```

iSazonov replied:
Now all properties is auto-property.

lzybkr replied:
Sorry, I don't see that change.

iSazonov replied:
My bad - lost commit.

Fixed.

lzybkr commented on src/System.Management.Automation/engine/lang/parserutils.cs:
```diff
@@ -1521,6 +1615,81 @@ public bool MoveNext()
+    /// The simple enumerator class is used for the range operator '..'
+    /// in expressions like 'A'..'B' | ForEach-Object { $_ }
+    /// </summary>
+    internal class CharRangeEnumerator : IEnumerator
```
The code is basically the same as RangeEnumerator, I wonder if there is a clever way we could avoid the duplication.

One thought - you could maybe derive from RangeEnumerator and just override Current.

It's probably not worth spending time on this though - C# doesn't handle this scenario nicely like C++ template would.

iSazonov replied:
I started with this but it looked too confusing. In addition, I thought that if we wanted to enhance this operator in the future, it would be easier if each type would have its own enumerator. Also possible we can make optimization and reduce string allocations.

iSazonov committed:
Use C# 7.0 pattern and helper function for Range operator

Use C# 7.0 pattern and helper function for Range operator
Use auto-properties in Range operator enumerator.

lzybkr commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -292,6 +286,10 @@ internal static class CachedReflectionInfo
             typeof(ParserOps).GetMethod(nameof(ParserOps.MatchOperator), staticFlags);
+        internal static readonly MethodInfo ParserOps_RangeOperator =
+            typeof(ParserOps).GetMethod(nameof(ParserOps.RangeOperator), staticFlags);
+        internal static readonly MethodInfo ParserOps_RangeEnumerator =
```
There is a convention here - the field name is class_member - so this should be ParserOps_GetRangeEnumerator.

iSazonov replied:
Fixed.

iSazonov committed:
Use auto-properties in CharRangeEnumerator

iSazonov committed:
Replace ParserOps_RangeEnumerator with ParserOps_GetRangeEnumerator

lzybkr commented on src/System.Management.Automation/engine/lang/parserutils.cs:
```diff
@@ -1478,7 +1567,13 @@ internal int UpperBound
         private int _current;
-        public object Current
+
+        Object IEnumerator.Current
```
Can you explain why this method was added? It seems like we only need the explicit method impl or the other, but not both.

iSazonov replied:
I saw it in CoreFX implementations ( [sample](https://github.com/dotnet/corefxlab/blob/b5a998d2758b8ee968dad324f104c9310057eda3/src/System.IO.Pipelines.Text.Primitives/SplitEnumerator.cs#L30) ). I don't remember exactly - it seems I'm checking that it doesn't work at all without this method if we use enheritance and call by IEnumerator interface.

iSazonov replied:
@lzybkr What is your conclusion?

lzybkr replied:
The example you point to is needed to implement the strongly typed interface `IEnumerable<T>`. If you wanted to follow that pattern, that's fine, but you would replace the `CurrentValue` property with the `Current` property.

iSazonov replied:
Fixed.

lzybkr commented on src/System.Management.Automation/engine/lang/parserutils.cs:
```diff
@@ -1521,6 +1616,75 @@ public bool MoveNext()
+
+        public string Current
+        {
+            get { return new String((char)CurrentValue, 1); }
```
Why convert to string? I was thinking the output should be char.

iSazonov replied:
Yes, it is _Char_ enumerator. 

Fixed.

iSazonov committed:
Refactor CharRangeEnumerator

Return char not string

iSazonov commented:
@lzybkr Do we ready to merge?

lzybkr commented on src/System.Management.Automation/engine/lang/parserutils.cs:
```diff
@@ -785,6 +785,95 @@ internal static object JoinOperator(ExecutionContext context, IScriptExtent erro
+        // In special case:
+        //     [char]'0'..[char]'9'
+        //     "0".."9"
+        // return objects of [int] type.
```
This comment only applies to the later case, not the former, right?

I don't recall the discussion around this, but I have a feeling we'll regret going this way despite it being a breaking change.

iSazonov replied:
Yes, I removed ` [char]'0'..[char]'9'`

iSazonov committed:
Fix comment

iSazonov commented:
@daxian-dbw Could you please fast review before I merge?

daxian-dbw commented:
@iSazonov I will quickly go through the changes later this morning.

iSazonov commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -4286,21 +4281,23 @@ public object VisitForEachStatement(ForEachStatementAst forEachStatementAst)
                     Expression lhs = Compile(binaryExpr.Left);
                     Expression rhs = Compile(binaryExpr.Right);
+                    var l = lhs.Cast(typeof(object));
+                    var r = rhs.Cast(typeof(object));
```
Fixed.

iSazonov commented on src/System.Management.Automation/engine/lang/parserutils.cs:
```diff
@@ -785,6 +785,94 @@ internal static object JoinOperator(ExecutionContext context, IScriptExtent erro
+        /// <param name="lval">The object on which to start.</param>
+        /// <param name="rval">The object on which to stop.</param>
+        /// <returns>The array of objects.</returns>
+        internal static object RangeOperator(ExecutionContext context, IScriptExtent errorPosition, object lval, object rval)
```
All other help methods use the pattern.
I can remove if you request.

daxian-dbw replied:
> All other help methods use the pattern.

If the arguments are not used in those help methods, they should also be removed. But you don't need to do that in this PR. Just fix the method added in this PR.

iSazonov replied:
Fixed.

iSazonov committed:
Remove unneeded variables

daxian-dbw commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -4874,14 +4871,11 @@ public object VisitBinaryExpression(BinaryExpressionAst binaryExpressionAst)
-                    return Expression.Call(CachedReflectionInfo.IntOps_Range,
-                                           lhs.Convert(typeof(int)),
-                                           rhs.Convert(typeof(int)));
+                    // TODO: replace this with faster code
```
I saw that discussion. The idea should be captured in the comment.

iSazonov replied:
The comment added.

iSazonov committed:
Add comment and remove unneeded parameters

iSazonov committed:
Fix expression call

daxian-dbw committed:
[Feature] Update the comment a bit

daxian-dbw commented:
@iSazonov According to https://github.com/PowerShell/PowerShell/pull/5732#discussion_r158558312, it seems to me you have fixed the breaking change. Is there still a breaking change introduced by this PR?

iSazonov commented:
@daxian-dbw I modify the PR description: "Add support for special case "0".."9" - return [int] objects."

@lzybkr Have you any thoughts or we can merge?

daxian-dbw commented:
@iSazonov I updated PR description a bit to make the breaking change more clear.
When squash merging, please make sure add `[breaking-change]` in the commit description, and summarize the breaking change.

iSazonov commented:
@lzybkr @daxian-dbw Thanks for help and great comments!

daxian-dbw commented:
@iSazonov The `[breaking-change]` tag should be added to the body of the commit message instead of the title of it. Sorry that I didn't point out that specifically.

When squash-merging a breaking change PR in future, please do like the following:
![image](https://user-images.githubusercontent.com/127450/36620586-a6379adc-18a8-11e8-9432-0fad261ed3d8.png)
